syne_tune/optimizer/schedulers/searchers/kde_searcher.py

```diff
@@ -0,0 +1,275 @@
+            points_to_evaluate: Optional[List[Dict]] = None,
+            **kwargs
+    ):
+        super().__init__(configspace=configspace, metric=metric, points_to_evaluate=points_to_evaluate)
+        self.mode = mode
+        self.metric_name = metric
+        self.num_evaluations = 0
+        self.min_bandwidth = min_bandwidth
+        self.random_fraction = random_fraction
+        self.num_candidates = num_candidates
+        self.bandwidth_factor = bandwidth_factor
+        self.points_to_evaluate = points_to_evaluate
+        self.top_n_percent = top_n_percent
+        self.X = []
+        self.y = []
+        self.categorical_maps = {
+            k: {cat: i for i, cat in enumerate(v.categories)}
+            for k, v in configspace.items()
+            if isinstance(v, sp.Categorical)
+        }
+        self.inv_categorical_maps = {
+            hp: dict(zip(map.values(), map.keys())) for hp, map in self.categorical_maps.items()
+        }
+
+        self.good_kde = None
+        self.bad_kde = None
+
+        self.vartypes = []
+        for hname in self.configspace:
+            hp = self.configspace[hname]
+            if isinstance(hp, sp.Categorical):
+                self.vartypes.append(('u', len(hp.categories)))
+            if isinstance(hp, sp.Integer):
+                self.vartypes.append(('o', (hp.lower, hp.upper)))
+            if isinstance(hp, sp.Float):
+                self.vartypes.append(('c', 0))
+
+        self.num_min_data_points = len(self.vartypes) if num_min_data_points is None else num_min_data_points
+        assert self.num_min_data_points >= len(self.vartypes)
+
+    @staticmethod
+    def to_feature(configspace, config, categorical_maps):
+        def numerize(value, domain, categorical_map):
+            if isinstance(domain, sp.Categorical):
+                res = categorical_map[value] / len(domain)
+                return res
+            else:
+                return [(value - domain.lower) / (domain.upper - domain.lower)]
+
+        return np.hstack([
+            numerize(value=config[k], domain=v, categorical_map=categorical_maps.get(k, {}))
+            for k, v in configspace.items()
+            if isinstance(v, sp.Domain)
+        ])
+
+    @staticmethod
+    def from_feature(configspace, feature_vector, inv_categorical_maps):
+        def inv_numerize(values, domain, categorical_map):
+            if not isinstance(domain, sp.Domain):
+                # constant value
+                return domain
+            else:
+                if isinstance(domain, sp.Categorical):
+                    index = int(values * len(domain))
+                    return categorical_map[index]
+                else:
+                    if isinstance(domain, sp.Float):
+                        return values * (domain.upper - domain.lower) + domain.lower
+                    else:
+                        return int(values * (domain.upper - domain.lower) + domain.lower)
```

Comment:
Maybe round here?

Reply:
Hmm shouldn't it be np.ceil, if I do the conversion that you described above?

Reply:
You need to decide whether it is worth adoption HyperparameterRanges. If so, I'd just use that code. But it is some extra work.